modules/phase_field/src/ics/PolycrystalVoronoiCoupledVoidIC.C

```diff
@@ -0,0 +1,104 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PolycrystalVoronoiCoupledVoidIC.h"
+
+// MOOSE includes
+#include "MooseMesh.h"
+#include "MooseVariable.h"
+#include "DelimitedFileReader.h"
+#include "GrainTrackerInterface.h"
+#include "PolycrystalVoronoi.h"
+
+InputParameters
+PolycrystalVoronoiCoupledVoidIC::actionParameters()
+{
+  InputParameters params = ::validParams<InitialCondition>();
+  params.addRequiredParam<unsigned int>("op_num", "Number of order parameters of grain");
+  params.addParam<bool>(
+      "columnar_3D", false, "3D microstructure will be columnar in the z-direction?");
+  params.addCoupledVar("v", "Coupled variable associated with the void");
+  params.addRequiredParam<Real>("invalue", "Variable value inside the coupled void");
+  params.addRequiredParam<Real>("outvalue", "Variable value outside the coupled void");
+  return params;
+}
+
+registerMooseObject("PhaseFieldApp", PolycrystalVoronoiCoupledVoidIC);
+
+InputParameters
+PolycrystalVoronoiCoupledVoidIC::validParams()
+{
+  InputParameters params = PolycrystalVoronoiCoupledVoidIC::actionParameters();
+  params.addClassDescription("Creates polycrystal voronoi grain structure aroud"
+                             "void positions obtained from the coupled variable");
+  params.addParam<unsigned int>("op_index", 0, "The index for the current order parameter");
+  params.addRequiredParam<UserObjectName>(
+      "polycrystal_ic_uo", "UserObject for obtaining the polycrystal grain structure.");
+  params.addParam<FileName>(
+      "file_name",
+      "",
+      "File containing grain centroids, if file_name is provided, the centroids "
+      "from the file will be used.");
+  return params;
+}
+
+PolycrystalVoronoiCoupledVoidIC::PolycrystalVoronoiCoupledVoidIC(const InputParameters & parameters)
+  : InitialCondition(parameters),
+    _op_num(getParam<unsigned int>("op_num")),
+    _op_index(getParam<unsigned int>("op_index")),
+    _columnar_3D(getParam<bool>("columnar_3D")),
+    _poly_ic_uo(getUserObject<PolycrystalVoronoi>("polycrystal_ic_uo")),
+    _file_name(getParam<FileName>("file_name")),
+    _var_val(coupledValue("v")),
+    _invalue(getParam<Real>("invalue")),
+    _outvalue(getParam<Real>("outvalue"))
+{
+  if (_invalue < _outvalue)
+    mooseError("PolycrystalVoronoiCoupledVoidIC requires that the voids be "
+               "represented with invalue > outvalue");
+}
+
+void
+PolycrystalVoronoiCoupledVoidIC::initialSetup()
+{
+  if (_op_num <= _op_index)
+    mooseError("op_index is too large");
+
+  // Obtain total number and centerpoints of the grains
+  _grain_num = _poly_ic_uo.getNumGrains();
+  _centerpoints = _poly_ic_uo.getGrainCenters();
+
+  // Call initial setup from InitialCondition
+  InitialCondition::initialSetup();
+}
+
+Real
+PolycrystalVoronoiCoupledVoidIC::value(const Point & p)
```

Comment:
Lots of code duplication with `PolycrystalVoronoiVoidIC` (which seems like a mess itself - that `file_name` parameter does not seem to be used, and it shouldn't).

The right thing to do here would be to create an abstract base class with a `computeCircleCenters()` method that would do the random placement on GBs in `PolycrystalVoronoiVoidIC` and reading in circle parameters from a file in `PolycrystalVoronoiCoupledVoidIC`.

Reply:
Actually, no

Reply:
The right right thing would be to create an IC that _just_ places the voids on the grain boundaries defined by the voronoi user object! And then delete `PolycrystalVoronoiVoidIC` entirely and replace every use by `PolycrystalVoronoiCoupledVoidIC`.

You also add a `file_name` parameter in your class that does _nothing_.